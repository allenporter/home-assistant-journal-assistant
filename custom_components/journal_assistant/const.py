"""Constants for journal_assistant."""

DOMAIN = "journal_assistant"

VISION_MODEL_NAME = "gemini-1.5-flash"
CONF_API_KEY = "api_key"
CONF_MEDIA_SOURCE = "media_source"
CONF_NOTES = "notes"
DEFAULT_NOTES = [
    "Daily",
    "Weekly",
    "Monthly",
]
DEFAULT_NOTE_NAME = "Journal"

CONF_CONFIG_ENTRY_ID = "config_entry_id"
