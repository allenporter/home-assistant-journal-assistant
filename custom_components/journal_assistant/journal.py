"""Converter from yaml journal files to an RFC5545 Journal."""

from pathlib import Path
import datetime
import logging

from ical.calendar import Calendar
from ical.journal import Journal
from .model import JournalPage


_LOGGER = logging.getLogger(__name__)


def journal_pages(storage_dir: Path, journal_name: str) -> list[JournalPage]:
    """Load all journal pages from a storage directory with the specified journal prefix."""
    pages = []
    files = list(storage_dir.glob(f"{journal_name}-*.yaml"))
    files.sort()
    for filename in files:
        with filename.open() as file:
            page = JournalPage.from_yaml(file.read())
        pages.append(page)
    return pages


def get_dated_content(page: JournalPage) -> dict[str, list[str]]:
    """Get the date and content from a journal page."""
    default_date = page.date or page.created_at
    if not page.records:
        return {default_date: [str(page.content)]}

    dated_content: dict[str, list[str]] = {}
    for note in page.records:
        note_date = note.date or default_date
        if note_date not in dated_content:
            dated_content[note_date] = []
        dated_content[note_date].append(f"- {note.content}")
    return dated_content


def journal_from_yaml(storage_dir: Path) -> dict[str, Calendar]:
    """Convert a yaml journal to an RFC5545 Journal."""
    _LOGGER.debug("Loading journal content from %s", storage_dir)
    # Get the unique list of journal names
    journal_names = {
        filename.name.split("-")[0] for filename in storage_dir.glob("*.yaml")
    }
    _LOGGER.debug("Journal names: %s", journal_names)

    journals = {}

    keys = list(journal_names)
    keys.sort()
    for journal_name in keys:

        # Load all pages from with the same journal prefix
        pages = journal_pages(storage_dir, journal_name)

        dated_content: dict[str, list[str]] = {}
        for page in pages:
            dated_content.update(get_dated_content(page))

        if journal_name not in journals:
            journals[journal_name] = Calendar()

        # Add a journal entry for each date
        calendar = journals[journal_name]
        for date, content_list in dated_content.items():
            journal = Journal()
            journal.summary = f"{journal_name} {date}"
            if "T" in date:
                journal.dtstart = datetime.datetime.fromisoformat(date)
            else:
                journal.dtstart = datetime.date.fromisoformat(date)
            journal.description = "\n".join(content_list)
            calendar.journal.append(journal)

        journals[journal_name] = calendar

    return journals
