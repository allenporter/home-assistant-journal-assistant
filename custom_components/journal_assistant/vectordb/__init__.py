"""Vector DB component for Home Assistant."""

from dataclasses import dataclass, field
from abc import ABC, abstractmethod
from typing import Any
import datetime
from collections.abc import Callable, Awaitable

from mashumaro.mixins.json import DataClassJSONMixin
from mashumaro.config import BaseConfig
import numpy as np


class VectorDBError(Exception):
    """Base class for VectorDB errors."""


@dataclass(kw_only=True)
class Embedding:
    """An embedding."""

    embedding: np.ndarray
    """The embedding."""


EmbeddingFunction = Callable[[list[str]], Awaitable[list[Embedding]]]
"""A function that takes a string and returns an embedding."""


@dataclass(kw_only=True)
class IndexableDocument(DataClassJSONMixin):
    """An indexable document."""

    uid: str
    """Unique identifier for the document."""

    timestamp: datetime.datetime | None
    """Timestamp of the document used for date restricts."""

    metadata: dict[str, Any] = field(default_factory=dict)
    """Metadata fields the document that can be searched."""

    document: str
    """The document content that will be indexed."""

    class Config(BaseConfig):
        omit_none = False
        code_generation_options = ["TO_DICT_ADD_OMIT_NONE_FLAG"]


@dataclass(kw_only=True)
class QueryParams(DataClassJSONMixin):
    """Query parameters for the VectorDB."""

    query: str | None = None
    """The query string."""

    start_date: datetime.datetime | None = None
    """Only include document chunks on or after this date."""

    end_date: datetime.datetime | None = None
    """Only include document chunks on or before this date."""

    metadata: dict[str, Any] | None = None

    num_results: int | None = None
    """Maximum number of results to return."""

    class Config(BaseConfig):
        omit_none = False
        code_generation_options = ["TO_DICT_ADD_OMIT_NONE_FLAG"]


@dataclass(kw_only=True)
class QueryResult(DataClassJSONMixin):
    """Query result from the VectorDB."""

    document: IndexableDocument
    """The document that was found."""

    score: float
    """The similarity score of the document."""


class VectorDB(ABC):
    """Journal Assistant vector search database."""

    @abstractmethod
    async def upsert_index(self, documents: list[IndexableDocument]) -> None:
        """Add notebooks to the index."""

    @abstractmethod
    async def count(self) -> int:
        """Return the number of documents in the collection."""

    @abstractmethod
    async def query(self, params: QueryParams) -> list[QueryResult]:
        """Search the VectorDB for relevant documents."""
