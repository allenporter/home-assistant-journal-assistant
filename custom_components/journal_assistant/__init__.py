"""journal_assistant custom component."""

from __future__ import annotations

import logging

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant

from .const import DOMAIN
from .services import async_register_services
from .llm import async_register_llm_apis
from .types import JournalAssistantConfigEntry
from .storage import create_vector_db

__all__ = [
    "DOMAIN",
]

_LOGGER = logging.getLogger(__name__)


PLATFORMS = (Platform.CALENDAR, Platform.SENSOR)


async def async_setup_entry(
    hass: HomeAssistant, entry: JournalAssistantConfigEntry
) -> bool:
    """Set up a config entry."""
    entry.runtime_data = await create_vector_db(hass, entry)
    await hass.config_entries.async_forward_entry_setups(
        entry,
        platforms=PLATFORMS,
    )
    async_register_services(hass)
    await async_register_llm_apis(hass, entry)
    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(
        entry,
        PLATFORMS,
    )
