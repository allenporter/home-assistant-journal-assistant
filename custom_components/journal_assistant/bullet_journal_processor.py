"""Processes a media source as a Bullet Journal.

This module is used to process a media source as a Bullet Journal. It extracts
text from the media source using a vision model and updates the vector database
with new journal entries.
"""

from typing import Any
import logging

from homeassistant.core import HomeAssistant, CALLBACK_TYPE, Event

from .media_source_listener import async_create_media_source_listener
from .const import DOMAIN

_LOGGER = logging.getLogger(__name__)


DATA_KEY = "bullet_journal_processor"


def async_register(
    hass: HomeAssistant, config_entry_id: str, media_source_domain: str
) -> CALLBACK_TYPE:
    """Register the media source listener."""
    listener_id = f"{DOMAIN}_{config_entry_id}"

    processor = BulletJournalProcessor(hass)
    cancel = hass.bus.async_listen(listener_id, processor.handle_event)
    unsub = async_create_media_source_listener(hass, media_source_domain, listener_id)

    def cleanup() -> None:
        """Clean up the media source listener."""
        cancel()
        unsub()

    return cleanup


class BulletJournalProcessor:
    """Library for processing a media source as a Bullet Journal."""

    def __init__(self, hass: HomeAssistant) -> None:
        """Initialize the Bullet Journal Processor."""
        self._hass = hass

    async def handle_event(self, event: Event[dict[str, Any]]) -> None:
        """Handle an updated media source object."""
        _LOGGER.info("Processing media source event: %s", event)
