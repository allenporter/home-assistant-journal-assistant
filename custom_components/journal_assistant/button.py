"""Journal assistant button platform."""

import datetime
import logging


from homeassistant.const import EntityCategory
from homeassistant.core import HomeAssistant
from homeassistant.components.button import ButtonEntity
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import DOMAIN
from .types import JournalAssistantConfigEntry

_LOGGER = logging.getLogger(__name__)

UPDATE_INTERVAL = datetime.timedelta(minutes=15)


async def async_setup_entry(
    hass: HomeAssistant,
    entry: JournalAssistantConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the journal sensor component."""
    async_add_entities([ProcessMediaButtonEntity(entry)], True)


class ProcessMediaButtonEntity(ButtonEntity):
    """A button that allows manual processing of media."""

    _attr_has_entity_name = True
    _attr_entity_category = EntityCategory.DIAGNOSTIC

    def __init__(self, entry: JournalAssistantConfigEntry) -> None:
        """Initialize the process media button entity."""
        self._attr_unique_id = f"{entry.entry_id}-process-media"
        self._processor = entry.runtime_data.media_source_processor
        self._attr_name = "Process Media"
        self._attr_device_info = {
            "identifiers": {(DOMAIN, entry.entry_id)},
            "name": entry.title,
        }

    async def async_press(self) -> None:
        """Handle the button press."""
        self.hass.async_create_task(
            self._processor.async_process_media(datetime.datetime.now())
        )
