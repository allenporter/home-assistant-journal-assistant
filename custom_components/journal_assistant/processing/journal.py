"""Converter from yaml journal files to an RFC5545 Journal."""

from pathlib import Path
import datetime
import logging
import hashlib
from typing import cast

from ical.calendar import Calendar
from ical.journal import Journal

from .model import JournalPage

_LOGGER = logging.getLogger(__name__)

__all__ = [
    "journal_from_yaml",
]


def journal_pages(storage_dir: Path, journal_name: str) -> list[JournalPage]:
    """Load all journal pages from a storage directory with the specified journal prefix."""
    pages = []
    files = list(storage_dir.glob(f"{journal_name}-*.yaml"))
    files.sort()
    for filename in files:
        with filename.open() as file:
            page = JournalPage.from_yaml(file.read())
        pages.append(page)
    return pages


def get_dated_content(page: JournalPage) -> dict[str, list[str]]:
    """Get the date and content from a journal page."""
    default_date = page.date or page.created_at
    if not page.records:
        return {default_date: [str(page.content)]}

    dated_content: dict[str, list[str]] = {}
    for note in page.records:
        note_date = note.date or default_date
        if note_date not in dated_content:
            dated_content[note_date] = []
        content = ""
        if note.content:
            if note.status:
                content = f"- ({note.status}) {note.content}"
            else:
                content = f"- {note.content}"
        if note.entries:
            if content:
                content += "\n"
            content += "\n".join(f"  - {entry}" for entry in note.entries if entry)
        dated_content[note_date].append(content)
    return dated_content


def write_content(content: str, filename: Path) -> None:
    """Write content to a file."""
    with filename.open("w") as file:
        file.write(content)


def write_journal_page_yaml(
    storage_dir: Path,
    note_name: str,
    page: JournalPage,
) -> None:
    """Write a journal page yaml string to a yaml file."""
    if "-" not in note_name:
        raise ValueError(f"Note name must contain a '-' character: {note_name}")
    if not page.filename.startswith(note_name):
        raise ValueError(
            f"Note name must match page filename: {note_name} != {page.filename}"
        )
    storage_dir.mkdir(parents=True, exist_ok=True)
    filename = storage_dir / f"{note_name}.yaml"
    _LOGGER.debug("Writing journal page to %s", filename)
    content = cast(str, page.to_yaml())
    write_content(content, filename)


def journal_from_yaml(
    storage_dir: Path,
    allowed_notes: set[str],
    default_note_name: str,
) -> dict[str, Calendar]:
    """Convert a yaml journal to an RFC5545 Journal."""
    _LOGGER.debug("Loading journal content from %s", storage_dir)
    # Get the unique list of journal names
    filename_prefixes = {
        filename.name.split("-")[0] for filename in storage_dir.glob("*.yaml")
    }
    note_names = list(filename_prefixes)
    note_names.sort()
    _LOGGER.debug("Journal names: %s", note_names)

    journals = {}
    for note_name in note_names:

        # Load all pages from with the same journal prefix
        pages = journal_pages(storage_dir, note_name)

        # Allow notes to have their own calendar entry if in the list of allowed notes
        key_name = note_name if note_name in allowed_notes else default_note_name

        dated_content: dict[str, list[str]] = {}
        for page in pages:
            for date, content_list in get_dated_content(page).items():
                if date not in dated_content:
                    dated_content[date] = []
                dated_content[date].extend(content_list)

        if key_name not in journals:
            journals[key_name] = Calendar()

        # Add a journal entry for each date
        calendar = journals[key_name]
        for date, content_list in dated_content.items():
            journal = Journal()
            journal.uid = hashlib.sha256(f"{note_name}-{date}".encode()).hexdigest()
            journal.summary = f"{note_name} {date}"
            journal.categories = [key_name]
            if note_name not in allowed_notes:
                journal.categories.append(note_name)
            if "T" in date:
                journal.dtstart = datetime.datetime.fromisoformat(date)
            else:
                journal.dtstart = datetime.date.fromisoformat(date)
            journal.description = "\n".join(content_list)
            calendar.journal.append(journal)

        journals[key_name] = calendar

    return journals
