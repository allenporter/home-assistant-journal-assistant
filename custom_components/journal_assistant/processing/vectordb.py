"""Journal Assistant vector search database."""

import itertools
import logging
from typing import Any
from pathlib import Path
import yaml


import chromadb
from chromadb.api.types import IncludeEnum
from chromadb.config import DEFAULT_TENANT, DEFAULT_DATABASE, Settings
from chromadb.utils.embedding_functions import google_embedding_function

from ical.calendar import Calendar
from ical.journal import Journal


_LOGGER = logging.getLogger(__name__)

BATCH_SIZE = 10
COLLECTION_NAME = "journal_assistant"
MODEL = "models/text-embedding-004"


def serialize_content(item: Journal) -> str:
    """Serialize a journal entry."""
    return yaml.dump(item.dict(exclude={"uid", "dtsamp"}, exclude_unset=True, exclude_none=True))  # type: ignore[no-any-return]


class VectorDB:
    """Journal Assistant vector search database."""

    def __init__(self, storage_path: Path, google_api_key: str) -> None:
        """Initialize the vector database."""
        _LOGGER.debug("Creating ChromaDB System")
        settings = Settings(
            anonymized_telemetry=False,
        )
        self.system = chromadb.config.System(settings)
        _LOGGER.debug("Creating Google embedding function")
        # Separate embedding functions are used for idnexing vs querying
        self.index_embedding_function = (
            google_embedding_function.GoogleGenerativeAiEmbeddingFunction(
                api_key=google_api_key, model_name=MODEL, task_type="RETRIEVAL_DOCUMENT"
            )
        )
        self.query_embedding_function = (
            google_embedding_function.GoogleGenerativeAiEmbeddingFunction(
                api_key=google_api_key, model_name=MODEL, task_type="RETRIEVAL_QUERY"
            )
        )
        self.client = chromadb.PersistentClient(
            path=str(storage_path),
            settings=settings,
            tenant=DEFAULT_TENANT,
            database=DEFAULT_DATABASE,
        )
        _LOGGER.debug("Creating collection: %s", COLLECTION_NAME)

    def upsert_index(self, notebooks: dict[str, Calendar]) -> None:
        """Add notebooks to the index."""
        _LOGGER.debug("Adding notebooks to index")

        collection = self.client.get_or_create_collection(
            name=COLLECTION_NAME, embedding_function=self.index_embedding_function  # type: ignore[arg-type]
        )

        for note_name, calendar in notebooks.items():
            _LOGGER.debug(
                "Indexing %s with %s entries", note_name, len(calendar.journal)
            )
            for found_journal_entries in itertools.batched(
                calendar.journal, BATCH_SIZE
            ):
                _LOGGER.debug(
                    "Examining batch of %s to index", len(found_journal_entries)
                )
                ids = [
                    journal_entry.uid or "" for journal_entry in found_journal_entries
                ]
                results = collection.get(ids=ids, include=[IncludeEnum.documents])
                _LOGGER.debug("Results: %s", results)
                existing_ids = {
                    uid: documents
                    for uid, documents in zip(
                        results["ids"], results.get("documents") or []
                    )
                }
                if existing_ids:
                    _LOGGER.debug("Found %s existing documents", len(existing_ids))
                # Determine which journal entries are entirely new or have
                # updated descriptions
                journal_entries = []
                documents = []
                metadatas = []
                for journal_entry in found_journal_entries:
                    existing_content = existing_ids.get(journal_entry.uid)
                    entry_content = serialize_content(journal_entry)
                    if existing_content is None or entry_content != existing_content:
                        journal_entries.append(journal_entry)
                        documents.append(entry_content)
                        metadatas.append(
                            {
                                "category": (
                                    journal_entry.categories[0]
                                    if journal_entry.categories
                                    else ""
                                ),
                                "date": journal_entry.dtstart.isoformat(),
                                "name": journal_entry.summary or "",
                            }
                        )
                ids = [journal_entry.uid or "" for journal_entry in journal_entries]
                if not ids:
                    _LOGGER.debug("Skipping batch of unchanged documents")
                    continue
                _LOGGER.debug("Upserting batch of %s to index", len(journal_entries))
                collection.upsert(
                    documents=documents,
                    metadatas=metadatas,  # type: ignore[arg-type]
                    ids=ids,
                )

    def count(self) -> int:
        """Return the number of documents in the collection."""
        collection = self.client.get_collection(
            name=COLLECTION_NAME, embedding_function=self.query_embedding_function  # type: ignore[arg-type]
        )
        return collection.count()

    def query(self, query: str, num_results: int) -> list[dict[str, Any]]:
        """Search the VectorDB for relevant documents."""
        collection = self.client.get_collection(
            name=COLLECTION_NAME, embedding_function=self.query_embedding_function  # type: ignore[arg-type]
        )

        results: chromadb.QueryResult = collection.query(
            query_texts=query,
            n_results=num_results,
        )
        if (
            (id_list := results.get("ids")) is None
            or (metadata_list := results.get("metadatas")) is None
            or (document_list := results.get("documents")) is None
        ):
            raise ValueError(f"Invalid query results: {results}")
        # Get the results for the first (and only) query
        ids = id_list[0]
        metadatas = metadata_list[0]
        documents = document_list[0]
        return [
            {
                "id": ids[index],
                "content": documents[index],
                **metadatas[index],
            }
            for index in range(len(ids))
        ]
