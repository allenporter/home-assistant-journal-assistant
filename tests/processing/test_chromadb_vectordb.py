"""Test loading the vector DB."""

from pathlib import Path
from collections.abc import Generator
from typing import Any
from unittest.mock import patch
import hashlib
import tempfile
import datetime

import chromadb
import numpy as np
from syrupy import SnapshotAssertion
import pytest

from custom_components.journal_assistant.const import DOMAIN
from custom_components.journal_assistant.processing import chromadb_vectordb
from custom_components.journal_assistant.processing.journal import journal_from_yaml
from custom_components.journal_assistant.vectordb import (
    QueryParams,
)


@pytest.fixture(autouse=True)
def mock_disable_posthog() -> Generator[None, None, None]:
    """Disable chromadb posthog since it starts daemon threads during tests."""
    with patch(
        "chromadb.telemetry.product.posthog.Posthog",
        autospec=True,
    ):
        yield


@pytest.fixture(name="storage_path")
def mock_vectordb_storage_path() -> Generator[Path, None, None]:
    """Fake out the storage path to load from the fixtures directory."""
    with tempfile.TemporaryDirectory() as tmpdir:
        yield Path(tmpdir)


class FakeEmbeddingFunction(chromadb.EmbeddingFunction):
    """Fake embedding function for testing."""

    embeds: int = 0

    def __call__(self, input: Any) -> chromadb.Embeddings:
        result: chromadb.Embeddings = []
        for item in input:
            self.embeds += 1
            result.append(
                np.array(
                    [ord(c) for c in hashlib.sha256(item.encode()).hexdigest()][0:3]
                )
            )
        return result


@pytest.fixture(name="embedding_function")
def mock_embedding_function() -> Generator[FakeEmbeddingFunction, None, None]:
    """Fixture to mock the embedding function."""
    fake_embedding = FakeEmbeddingFunction()
    with patch(
        f"custom_components.{DOMAIN}.processing.chromadb_vectordb.google_embedding_function.GoogleGenerativeAiEmbeddingFunction",
        return_value=fake_embedding,
    ):
        yield fake_embedding


def test_vectordb_loading(
    storage_path: Path,
    embedding_function: FakeEmbeddingFunction,
    snapshot: SnapshotAssertion,
) -> None:
    """Test parsing a journal page."""

    entries = journal_from_yaml(Path("tests/fixtures"), {"Daily", "Monthly"}, "Journal")
    assert len(entries) == 3
    assert entries.keys() == {"Daily", "Journal", "Monthly"}

    # Add the first entry to the index
    first_calendar = next(iter(entries.values()))
    client = chromadb_vectordb.create_local_chroma_client(storage_path)
    db = chromadb_vectordb.ChromaVectorDB(client, "12345")
    db.upsert_index(
        [
            chromadb_vectordb.create_indexable_document(entry)
            for entry in first_calendar.journal
        ]
    )
    assert embedding_function.embeds == 4

    # Add the rest, which skips the duplicate
    db.upsert_index(
        [
            chromadb_vectordb.create_indexable_document(entry)
            for calendar in entries.values()
            for entry in calendar.journal
        ]
    )
    assert embedding_function.embeds == 7

    assert db.count() == 7

    results = db.query(QueryParams(query="tree", num_results=5))
    assert len(results) == 5
    assert results[0] == snapshot

    # Limit results to a single date
    results = db.query(
        QueryParams(
            query="gifts",
            category="Daily",
            start_date=datetime.date(2023, 12, 21),
            end_date=datetime.date(2023, 12, 21),
            num_results=5,
        )
    )
    assert len(results) == 1
    assert results[0] == snapshot
