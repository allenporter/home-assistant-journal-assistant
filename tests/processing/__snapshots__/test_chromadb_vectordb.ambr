# serializer version: 1
# name: test_vectordb_loading
  QueryResult(document=IndexableDocument(uid='3ac14d0bb8c28ac12733b156a5b96d6af62d3c3be63d488aa763a99637a407ce', timestamp=datetime.datetime(2023, 12, 22, 6, 0), metadata={'category': 'Daily', 'date': 1703224800.0, 'name': 'Daily 2023-12-22'}, document="categories:\n- Daily\ndescription: '- clean garage'\ndtstart: 2023-12-22\nsummary: Daily 2023-12-22\n"), score=5.000000000000001)
# ---
# name: test_vectordb_loading.1
  QueryResult(document=IndexableDocument(uid='caba4b5990a778e89764bdb09f6902d6fc68a48d8fe96c1d9bbe4424d00af930', timestamp=datetime.datetime(2023, 12, 21, 6, 0), metadata={'category': 'Daily', 'date': 1703138400.0, 'name': 'Daily 2023-12-21'}, document="categories:\n- Daily\ndescription: '- cardboard breakdown\n\n  - (migrated) Bowling w/ Q\n\n  - (completed) flux-local helm\n\n  - todo urls?\n\n  - (migrated) gifts plan\n\n  - windows xmas lights\n\n  - (migrated) fitbit python'\ndtstart: 2023-12-21\nsummary: Daily 2023-12-21\n"), score=2541.0)
# ---
